fix storing id_array files when graph is moved

e-mcginnis opened the pull request:
When an RDG is moved, the topology, node_id_array, and edge_id_array
files must also be moved
previously, this was determined by checking if the headers path to these
files was empty. This relied on ChangeStorageLocation() actually
emptying them.

- make ChangeStorageLocation empty the node/edge id_array paths
- store the topology, node/edge id_array files if the rdg path
  no longer matches the old rdg path

- add node/edge id array files to RDGManifest::FileNames
- add handful of debug log lines to the store/write path
- explicitly log when EntityTypeIds are being assigned from
  properties
- we also would end up storing the id_array files twice, remove second
  place

this bug was exercised when:
1) take a graph that already has EntityTypeID structures
2) load it
3) store it in a new location
4) load from the new location

our existing tests only do:
1) take a graph *without* EntityTypeID structures
2) load it
3) store it in a new location, which stores the new EntityTypeID structures
4) load from the new location

so this path got missed. Will push a future PR with a new test for this path.

bowu commented on libquery/src/CaseExpression.cpp:
```diff
@@ -0,0 +1,107 @@
+  Column default_col = default_exp_->Evaluate(table, graph);
+
+  std::vector<int> ind_array;
+  switch (val_col.GetType()) {
```
@danielmawhirter Now that we have the Column class, could you think about how to merge Column and Table in some better way to simplify the code elsewhere?

danielmawhirter replied:
Yes, I'll work on that next week

ddn0 commented on libquery/include/querying/Expression.h:
```diff
@@ -82,6 +82,7 @@ enum ExpressionOperator {
 //! TODO(l-hoang) probably makes more sense to call it a literal tree
 class Expression {
 public:
+  Expression() = default;
```
do we really want to make an expression default constructable? What does `Expression x;` supposed to mean?

bowu replied:
This is required by the CaseExpression constructor. I'd appreciate your suggestion on this. @ddn0

ddn0 commented on libquery/include/querying/CaseExpression.h:
```diff
@@ -0,0 +1,39 @@
+      std::pair<std::unique_ptr<Expression>, std::unique_ptr<Expression>>;
+
+  CaseExpression(std::unique_ptr<Expression> case_exp,
+                 std::vector<CasePair>&& case_pairs,
```
Just write this as `std::vector<CasePair> case_pairs`; callers have to use `std::move` anyway because of unique_ptrs.

ddn0 commented on libquery/include/querying/CaseExpression.h:
```diff
@@ -0,0 +1,39 @@
+        default_exp_(std::move(default_exp)) {}
+  virtual ~CaseExpression() = default;
+
+  Column Evaluate(const Table* table,
```
why not take a `const&` rather than `const*`?

bowu replied:
I changed const& back to const* because otherwise filter-op-test cannot pass. We'd like to use an empty graph for unit testing, but we don't have an empty AttributedGraph constructor. Using const& leads to dereferencing the nullptr.

ddn0 replied:
Seems like the solution would be to just create an empty AttributedGraph rather than forcing the downstream code to deal with incompletely constructed inputs.

ddn0 commented on libquery/src/CypherCompiler.cpp:
```diff
@@ -703,11 +704,38 @@ galois::CypherCompiler::CompileExpression(CypherASTNode ast) {
+  }
+  std::unique_ptr<Expression> default_exp =
+      CompileExpression(cypher_ast_case_get_default(ast));
+  std::cout << "default_exp: " << default_exp->ToString() << "\n";
```
do we really want this outputted to stdout all the time?

bowu replied:
Will remove them.

ddn0 commented on libquery/include/querying/Column.h:
```diff
@@ -65,6 +80,32 @@ class Column {
 
+  template <typename T>
+  T GetPlainItem(int64_t i) const {
+    GALOIS_LOG_ASSERT(array_->length() > 0);
```
Properties.h is supposed to provide a uniform method for viewing arrow arrays.

bowu replied:
I believe Daniel will look into this and provide a nicer way to access table and column data next week.

ddn0 commented on libquery/include/querying/Column.h:
```diff
@@ -84,6 +125,21 @@ class Column {
+
+  template <typename T>
+  static Column Make(TableElementType type, const std::vector<T>& vec) {
+    using ArrowType   = typename arrow::CTypeTraits<T>::ArrowType;
```
how is this different than TableBuilder?

bowu replied:
It's not conceptually different, but easier to use when I only need to deal with columns. @danielmawhirter and I will revisit this after he integrates Column and Table.

ddn0 commented on libquery/include/querying/CaseExpression.h:
```diff
@@ -0,0 +1,39 @@
+public:
+  //(predicate, returned expression) pair
+  using CasePair =
+      std::pair<std::unique_ptr<Expression>, std::unique_ptr<Expression>>;
```
if you use a struct like:
```cpp
struct Case {
  std::unique_ptr<Expression> predicate;
  std::unique_ptr<Expression> result;
};
```

the code documents itself.

bowu replied:
Will do.

ddn0 commented on libquery/src/CaseExpression.cpp:
```diff
@@ -0,0 +1,107 @@
+                                        const std::vector<Column>& case_cols) {
+  std::vector<int> ind_array;
+  for (size_t i = 0; i < val_col.Size(); ++i) {
+    T val      = val_col.GetPlainItem<T>(i);
```
I thought we were trying to avoid element wise operations. Can we use an arrow::compute function here?

bowu replied:
I didn't find a good way to do it, but will give it another try.

bowu replied:
I don't think any arrow compute function or a composition of them can efficiently do what's needed here. I may be able to use some more complex logic to make it work awkwardly, but it would make the code hard to read. Basically, I have a column a, a vector of columns b, and a vector of columns c as input. If a[i] = b[j][i], output[i] = c[j][i], where b[j][i] means the ith element of the jth column in b. @ddn0 @danielmawhirter If you don't have suggestions, I'll leave it as is.

ddn0 replied:
I agree that there doesn't seem to be a if-else compute kernel exactly in arrow compute. The path is probably to define a new function in arrow compute via its registry API. Given a bunch of code in this PR exists to just add scalar support that arrow compute also has and that element-wise operations are generally slow, it might make sense to consider what path we want to take here.

ddn0 reviewed (commented):
can you add your test case to the test suite so we can some test coverage?

e-mcginnis committed:
fix storing id_array files when graph is moved

When an RDG is moved, the topology, node_id_array, and edge_id_array
files must also be moved
previously, this was determined by checking if the headers path to these
files was empty. This relied on ChangeStorageLocation() actually
emptying them.

- make ChangeStorageLocation empty the node/edge id_array paths
- store the topology, node/edge id_array files if the rdg path
  no longer matches the old rdg path

- add node/edge id array files to RDGManifest::FileNames
- add handful of debug log lines to the store/write path
- explicitly log when EntityTypeIds are being assigned from
  properties

- we also would end up storing the id_array files twice, remove second
  place

tylershunt reviewed (approved):
Nice

gliga reviewed (approved):
Great!

witchel commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -421,6 +421,7 @@ katana::PropertyGraph::DoWrite(
   std::unique_ptr<tsuba::FileFrame> edge_entity_type_id_array_res = nullptr;
 
   if (!rdg_.topology_file_storage().Valid()) {
+    KATANA_LOG_DEBUG("topology file store invalid, writing");
```
Evan, this comment is not specific to your PR, but about our logging levels.  @tylershunt I wonder if post-MVP we should discuss logging moving forward.  

LOG_DEBUG will print out for all debug builds, which generally isn't what I think most developers want.  They want a more "show me the details if I'm interested" level of logging.

witchel reviewed (commented):
Thanks, Evan.  While this fix looks like it takes care of the problem, I wonder about the next addition to the manifest structure.  Is there a way to improve the structure of the code to shift this type of error into compile time?

roshandathathri commented:
Thanks for the clear PR/commit message!

e-mcginnis commented:
> Thanks, Evan. While this fix looks like it takes care of the problem, I wonder about the next addition to the manifest structure. Is there a way to improve the structure of the code to shift this type of error into compile time?

the manifest structure itself is pretty straightforward to handle, its these other files (topology, id_arrays) that are more brittle. It would be nice to have a standard way to manage them. Its possible then that we could add some compile time checks to help here, but how to go about it is not immediately obvious to me.

witchel commented:
> > Thanks, Evan. While this fix looks like it takes care of the problem, I wonder about the next addition to the manifest structure. Is there a way to improve the structure of the code to shift this type of error into compile time?
> 
> the manifest structure itself is pretty straightforward to handle, its these other files (topology, id_arrays) that are more brittle. It would be nice to have a standard way to manage them. Its possible then that we could add some compile time checks to help here, but how to go about it is not immediately obvious to me.

I'm not sure.  The only thing I could think of is an enumerated type, one for each other file.  When you add a new file, you add a type, and then there is some switch statement somewhere that will fail until you add code to deal with that file's operations.
